src/cmap/wait_queue.rs

```diff
@@ -1,177 +1,85 @@
+    pub(super) fn new(address: StreamAddress, max_handles: u32, timeout: Option<Duration>) -> Self {
+        let max_handles = max_handles.try_into().unwrap_or(usize::max_value());
 
         Self {
-            inner: Arc::new(Mutex::new(inner)),
+            semaphore: Semaphore::new(true, max_handles),
+            address,
+            timeout,
+            max_handles,
         }
     }
 
-    /// Enters the wait queue and blocks until it either reaches the front of the queue or the
-    /// timeout has elapsed.
-    pub(super) fn wait_until_at_front(&self) -> Result<WaitQueueHandle> {
-        let mut guard = self.inner.lock().unwrap();
+    /// Enter the wait queue and block until either reaching the front of the queue or
+    /// exceeding the timeout.
+    pub(super) async fn wait_until_at_front(&self) -> Result<WaitQueueHandle<'_>> {
+        let future = self.semaphore.acquire(1);
 
-        let condvar = Arc::new(Condvar::new());
-        guard.queue.push_back(condvar.clone());
-
-        // If we're already at the front of the wait queue, we don't need to wait. We don't pop
-        // ourself off of the wait queue yet though, since another thread may enter while we're
-        // waiting for a connection to become available.
-        if guard.queue.len() == 1 {
-            return Ok(WaitQueueHandle::new(guard, condvar));
-        }
-
-        let guard = if let Some(timeout) = guard.timeout {
-            // Wait until all of the WaitQueueHandles in the queue in front of the current thread's
-            // are dropped or a timeout occurs.
-            let (guard, result) = condvar.wait_timeout(guard, timeout).unwrap();
-
-            if result.timed_out() {
-                return Err(ErrorKind::WaitQueueTimeoutError {
-                    address: guard.address.clone(),
+        let releaser = if let Some(timeout) = self.timeout {
+            RUNTIME.await_with_timeout(Box::pin(future), timeout).await.map_err(|_| {
+                ErrorKind::WaitQueueTimeoutError {
+                    address: self.address.clone(),
                 }
-                .into());
-            }
-
-            guard
+            })?
         } else {
-            // Wait until all of the WaitQueueHandles in the queue in front of the current thread's
-            // are dropped.
-            condvar.wait(guard).unwrap()
+            future.await
         };
 
-        Ok(WaitQueueHandle::new(guard, condvar))
+        Ok(WaitQueueHandle {
+            semaphore_releaser: releaser
+        })
     }
 
-    /// Notifies the wait queue that the next thread in the queue can be woken up.
-    pub(super) fn notify_ready(&self) {
-        self.inner.lock().unwrap().notify_ready();
+    /// Signal that the front of the queue (if there is one) is ready to wake up.
+    pub(super) fn wake_front(&self) {
+        if self.semaphore.permits() >= self.max_handles {
+            panic!("greater than {} connections checked back into pool with address {}",
```

Comment:
So far, we've generally been panicking when preconditions aren't met (e.g. `unwrap` on locks that could be poisoned), so I continued that here. We _could_ silently handle this here if we wanted to, though.

One thing that I'm noticing now that seems kind of troubling is that the semaphore [appears to be implemented using a parking lot mutex](https://docs.rs/futures-intrusive/0.3.0/futures_intrusive/sync/type.Semaphore.html), so that [`release()`](https://docs.rs/futures-intrusive/0.3.0/src/futures_intrusive/sync/semaphore.rs.html#524-526) and [`permits()`](https://docs.rs/futures-intrusive/0.3.0/src/futures_intrusive/sync/semaphore.rs.html#529-531) constitute blocking calls (note how this method isn't `async`). Should this be something we're worried about?

Reply:
I don't think that they are actually blocking; it looks like they write their own wrapper around the parking lot primitives using things like `try_lock` rather than `lock`; note that in the body of the code you linked to, `self.state` refers to a `PollState`, not a `Mutex`. From glancing over the code [here](https://github.com/Matthias247/futures-intrusive/blob/master/src/sync/mutex.rs), it doesn't seem like they're actually calling `lock` on a mutex directly anywhere.

Reply:
Ok cool.